### ISSUE TITLE
some gitg observations

### ISSUE THREAD

**Issue by ZachBacon** [Alexpux/MINGW-packages#942]
alright, love that gitg finally landed, however some small observations
I know based on some of the commits it isn't exactly ready but looking close. And yes the file does exist, I made sure I looked in the directory.

```
 $ gitg
C:/msys64/mingw64/bin/gitg.exe: error while loading shared libraries: libgitg-1.0-0.dll: cannot open shared object file: No such file or directory
```

**Comment by mingwandroid** [comment]
@DoctorWho11  What do you get if you double click on `gitg.exe` from `explorer`?

**Comment by DoctorWho11** [comment]
And I'm glad you mentioned that, missing libgee for some reason. Perhaps it needs to be added as a dep to the gitg package?

**Comment by mingwandroid** [comment]
It listed as a makedepends, but I guess it needs to be a depends, @nacho?

**Comment by nacho** [comment]
Yes this was going to be a change with the release i was thinking to make tomorrow

**Comment by mingwandroid** [comment]
No problem. I'm going to try to debug into the commit dialog thing, unless you are already on it?

**Comment by nacho** [comment]
I was not working on it, though i might have a patch in one of my branches. It is Just a matter of changing the pkgbuild file. If u get to it great, if not, I'll do it tomorrow.

**Comment by mingwandroid** [comment]
ok, I have a few minor updates to the pkgbuild that I will commit shortly:

options array removed (all were default)
libgee as depend
add pkgbase
pass --enable-debug if check_option "debug" "yes" (so it is debuggable in Qt Creator)

.. speaking of Qt Creator, do you think a good gnome-based IDE will be arriving some time? :-)

**Comment by DoctorWho11** [comment]
As long as things are getting there then its all good to me :)

**Comment by mingwandroid** [comment]
@nacho .. aaah
(gdb) b main
Breakpoint 1 at 0x4889d7: file /home/nacho/checkout/gnome/gitg/gitg/gitg.vala, line 27.
(gdb) r
Starting program: E:\msys64\mingw64\bin\gitg.exe
[New Thread 2664.0x15e0]

Breakpoint 1, main (argc=1, argv=0x2ae8e60) at /home/nacho/checkout/gnome/gitg/gitg/gitg.vala:27
27      /home/nacho/checkout/gnome/gitg/gitg/gitg.vala: No such file or directory.

/home/nacho/checkout ?!? OK, brave new (Vala) world for me then, I might give up now and file you that bug you asked for tomorrow instead! My push is incoming after I verify I didn't break anything.

**Comment by mingwandroid** [comment]
Pushed.

**Comment by DavidEGrayson** [comment]
> /home/nacho/checkout

pacman should be printing a warning when it sees a directory like that in your package, during the linting step.

**Comment by mingwandroid** [comment]
@DavidEGrayson this is in some embedded debugging info, so it is likely compressed in some manner. I wouldn't worry too much about the linting side of things. I'm a bit more surprised/concerned that I've rebuilt from source on my machine, yet this debugging info refers to a path on @nacho's machine; I guess valac didn't get run?

**Comment by mingwandroid** [comment]
anyway, the good thing to take away from all of this interest @nacho is that gitg looks so damn cool that we are all "chomping at the bit" to use it to help us to in our hacking ;-)

**Comment by nacho** [comment]
@mingwandroid feel free to file bugs, the point that it shows my path it is really weird, I thought that should be stripped away when distchecking. About a gnome IDE, even though I use gedit with some plugins (I am part of the maintainers) you should definitely have a look at the new IDE that is coming and that it is mostly complete now. gnome-builder: https://git.gnome.org/browse/gnome-builder

**Comment by nacho** [comment]
btw, I made a new release and pull requested an update for msys2. This contains some fixes for win32. Cheers

**Comment by nacho** [comment]
@mingwandroid Pointing you to this bug report https://github.com/libgit2/libgit2/pull/3557#issuecomment-166888568

See that fetch does not work on gitg for some reason coming from libgit2. I think our openssl is broken somehow. Would be cool if you could change the patch in that comment to instead do the printfs and report back what are the find outs on where exactly it fails. I will not be available for the next 15 days or so, so some help from your side will be appreciated.

**Comment by mingwandroid** [comment]
@nacho Ok great thanks for the info. Have a relaxing break.

**Comment by nacho** [comment]
@mingwandroid did you have time to research this a bit?

**Comment by mingwandroid** [comment]
Happy New Year @nacho, not really, I had some awful troubles with GEGL that @Alexpux asked me to look into which led to me digging quite deep into our Python 2.7.10 package. I updated it to 2.7.11 and added support for --with-pydebug so I can use gdb py-bt and the like.

I figured out the problem more or less - it was to do with distutils and some Windows argument quoting stuff (we can build GEGL and new gimp packages at least).

I also added glib2-git and gtk3-git packages that build with 'debug' and '!strip' options as I want to investigate these in that scenario. If you have time, I'd like to find (where applicable) the original sources for the patches at https://github.com/Alexpux/MINGW-packages/tree/master/mingw-w64-glib2-git so they can be applied with `git am` so that they contain and retain attribution and commit messages?

My job is about to get very busy again though, but I hope to be able to squeeze some time in for this, `gitg` and also `gnome-builder`.

**Comment by DoctorWho11** [comment]
I'll see about testing some more when I get home. I'll package up the newer gtk git packages as well. So far though current gitg build sees the Repo's but can't commit or clone.

**Comment by mingwandroid** [comment]
Hi @nacho, to force `valac` to be run, I had to do the following in the `prepare()` function in the `PKGBUILD`:

`find . -name "*.vala" -exec touch {} \;`

.. any ideas why that would be? Are the tarballs carrying vala object files incorrectly? What are the object files called even? Yeah, I'm very new to vala, but willing to learn.

**Comment by nacho** [comment]
@mingwandroid valac will generate the .c and .h files that are already included on the tarball. So not sure why you want to regenerate them with that. You should only generate them if you change something. Any reason for that?

**Comment by mingwandroid** [comment]
Yeah, so that gdb knows the right file when I enter gdb main.

**Comment by nacho** [comment]
@mingwandroid so this is what I've found in relation to http/s support. It seems the supported way for it on windows from libgit2 is to be using winhttp. There is a bug that should be fixed in master that avoids using openssl at all on windows and on our package we should force winhttp and disable openssl.


<cmn> there's absolutely no reason why we would want to use openssl on Windows, there are better things for doing TLS and HTTPS
<cmn> also, libssh2 on Windows does not depend on openssl
<cmn> so there's not even the issue about the thread initialisation
<cmn> which you shouldn't be depending on really
<cmn> there's no reason for either of these libraries to use openssl, it's superflous on Windows, we already have crypto on Windows

**Comment by mingwandroid** [comment]
@dscho you opinion please ..

Git for Windows uses OpenSSL, so therefore libgit2 on Windows should do the same.

**Comment by nacho** [comment]
@mingwandroid that chat pointed out before is from the libgit2 maintainer

**Comment by mingwandroid** [comment]
Sure. I need to gather the opinions of others too. @Alexpux?

**Comment by mingwandroid** [comment]
In the latest release of libgit2, it is still optional whether to use OpenSSL or winhttp, and we've elected not to use winhttp:
https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-libgit2/PKGBUILD#L39

**Comment by nacho** [comment]
@mingwandroid for what the developer told me is that openssl being available on windows is a bug that should be fixed in master. We should always use winhttp

**Comment by Alexpux** [comment]
Many software use OpenSSL on windows. Why not fix using it?

**Comment by nacho** [comment]
Well there are a few reasons:
 1. the maintainers prefer to support winhttp
 2. winhttp is provided by windows itself so I guess it is actually better than openssl

I agree though on the point that the current situation is not nice. Glib uses gnutls, other software openssl, other software schannel.

FWIW I tend to agree with the libgit2 maintainer that using the platform specific api is actually better as in this case is the platform the one dealing with security updates instead of the final application having to update openssl or gnutls or whatever.

PS: I have an openssl backend for glib in case you are interested.

**Comment by mingwandroid** [comment]
We want open-source to be strong on Windows, and for it to be strong means that open source projects need to be used by other open source projects rather than to avoid the same.

So I think we would prefer that MSYS2's build of libgit2 use OpenSSL. From the users' perspective, if our certificates are stored in different places for different aspects of the same "feature", in this case, using git from the command-line versus gitg, then it does not provide a good experience.

I'm sure there are good reasons from the perspective of Carlos Martín Nieto's and of the other Windows-centric libgit2 contributors to prefer it to use winhttp, but we need consensus. Either we switch everything over or we switch none of it, IMHO.

**Comment by mingwandroid** [comment]
I'm not sure I buy the argument that as winhttp is part of Windows it's better. I agree however that it's a guess, being as it's closed source ;-)

**Comment by nacho** [comment]
I do not say it is better I say that using the platform specific is better since it is the platform dealing with security updates for us :)

**Comment by nacho** [comment]
Also fyi https://github.com/wingtk/glib-networking/releases/tag/2.46.2-openssl

**Comment by nacho** [comment]
So I wonder if we should add this too to the package https://git.gnome.org/browse/gitg/commit/?id=fa1b007d9e766c8fa00fa69bbe263bd365c22d81

**Comment by mingwandroid** [comment]
Oh, yeah!

**Comment by dscho** [comment]
My perspective: OpenSSL has served us pretty well over the years, so just switching it out for WinHTTP (which is a complete unknown from Git for Windows' perspective) is most likely leading to problems. Only naive developers would expect everything to go smoothly with such a big switch.

Ideally, I would love to have a switch that tells cURL what to use: WinHTTP or OpenSSL?

As for Git for Windows: we require OpenSSL anyway, for its SHA-1 implementation. Of course, we could switch to another implementation, but you know, we have OpenSSL support, and it works, and I would be reluctant to switch to another implementation unless somebody I can trust does all the work (and I cannot be that person because I have too much on my plate already).

In summary: I understand libgit2's position, but just abandoning OpenSSL because of one developer's opinion would be a bit rash IMHO.

**Comment by carlosmn** [comment]
libgit2 tries to reduce the amount of dependencies to carry around, since being usable from bindings from other languages is very important for us, and the less you have to deal with C dependencies the better.

We also want to be a library that works *in* the platform where it's built, so that applications using this library can remain as integrated into the host OS as they would be without libgit2 making them or their users have to set up extra information just because we happen to write the library mostly on unix systems. This means we use CommonCrypto+SecureTransport on OSX and WinCNG+WinHTTP (we could use SChannel but WinHTTP gives us more functionality we can leverage) on Windows. For the rest we use OpenSSL for crypto as it's something you'd expect to find there. These libraries are managed by the OS packaging system so you don't have to perform extra updates or install extra software. With the exception of OpenSSL these are also thread-safe themselves and they don't make the final application developer/package have to care about making sure threading for second- or third-level dependency is set up correctly.

Drawing the line at WinHTTP vs OpenSSL seems rather arbitrary. We use *a lot* of Win32-specific functions to try to make up for the general slowness of metadata retrieval on NTFS. This is also functionality that's provided by the OS and it would hurt the users if we don't take advantage of them.

The same is true for the network actions. Not using WinHTTP means we don't share the certificate store with the rest of the system, and an application using such a version of libgit2 would be foreign to the user; this makes the application a second-class citizen, and I don't see that helping the FLOSS case on Windows.

Our approach is that if we're going to be on Windows, we're going to *be on Windows* and we should integrate as though we actually wrote the code on that OS (the first integrations with WinHTTP and ScureTransport code was written on a Debian machine).

The core team isn't going to support OpenSSL on Windows (I'm not quite sure why we even have that flag in CMake) but if a third party wants to contribute at patch, we'll take a look, provided it's not too disruptive for the initialisation code paths.

**Comment by carlosmn** [comment]
I also don't presume to tell git what to do. I would prefer it use the OSs own libraries (git can be built with CommonCrypto there, and the system-provided libcurl, which you'll generally link to, uses SecureTransport) but I know that there are a lot of unixisms in this tool, and some users do like that it comes with half a unix system. But libgit2 was made to avoid precisely that and build natively.

**Comment by mingwandroid** [comment]
@carlosmn It's good that you won't seek to actively remove Windows OpenSSL support, and on MSYS2 we'll continue to build and test our libgit2 packages that way since our git is (obviously) built using it and from our perspective consistency between the two is the most important concern here. We'll push patches to you for any problems we run into.

Thanks for your feedback.

**Comment by lazka** [comment]
Seems to me like everything here has been resolved, more or less. But feel free to reopen.

Slightly related, the openssl backend for glib which nacho mentioned is now in the repo, see #3487
